{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4332e670-8f43-4724-9e09-6478f3f419e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5ba36a96-f54c-43cb-92dd-7828f9d2b223",
   "metadata": {},
   "outputs": [],
   "source": [
    "def augustus_refs(df):\n",
    "    out_df= pd.DataFrame()\n",
    "    accessions=df[\"accession\"].unique().tolist()\n",
    "    for acc in accessions:    \n",
    "        busco = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/busco_{acc}.txt\", header=None)\n",
    "        house = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/housekeeping_{acc}.txt\", header=None)\n",
    "        nlr = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/nlrs_{acc}.txt\", header=None)\n",
    "\n",
    "        busco[\"type\"] = \"busco\"\n",
    "        house[\"type\"] = \"housekeeping\"\n",
    "        nlr[\"type\"] = \"nlr\"\n",
    "        gene_type = pd.concat([busco,house,nlr])\n",
    "        gene_type.columns = [\"ref_isoform\",\"type\"]\n",
    "        gene_type = gene_type.groupby(\"ref_isoform\")[\"type\"].apply(', '.join).reset_index()\n",
    "\n",
    "        gene_type[\"ref_gene\"] = gene_type[\"ref_isoform\"].str.split(\".\").str[0]\n",
    "        gene_type.loc[gene_type[\"type\"] == \"busco, nlr\"] = \"nlr\"\n",
    "        gene_type.loc[gene_type[\"type\"] == \"housekeeping, nlr\"] = \"nlr\"\n",
    "        gene_type.loc[gene_type[\"type\"] == \"busco, housekeeping\"] = \"housekeeping\"\n",
    "\n",
    "        sub=df.loc[df[\"accession\"]==acc]\n",
    "        sub = pd.merge(left=sub, right=gene_type, how='left', left_on='associated_gene', right_on='ref_gene')\n",
    "\n",
    "        \n",
    "        out_df=pd.concat([out_df,sub])\n",
    "    return out_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce4c345a-2540-4631-977d-90acd7a09eaa",
   "metadata": {},
   "source": [
    "## Access TAMA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d7eedf5d-cdd2-4861-bb9b-e957cd9a9ac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref=pd.read_pickle(\"all_runs/tama_augustus_sqanti.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c6f7413c-57f9-4518-961b-8567efeff6a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "full-splice_match          1760959\n",
       "novel_not_in_catalog        825013\n",
       "incomplete-splice_match     419668\n",
       "novel_in_catalog            206821\n",
       "genic                       166785\n",
       "antisense                    29343\n",
       "fusion                       28951\n",
       "intergenic                   22284\n",
       "genic_intron                     8\n",
       "Name: structural_category, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref[\"structural_category\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8e0cd82-ade9-4d6a-a5a8-b1388ed95bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load full df\n",
    "ref=pd.read_pickle(\"all_runs/tama_augustus_sqanti.pkl\")\n",
    "# add new column\n",
    "ref[\"cat_assign\"]=\"None\"\n",
    "# and fill new colum based on 4 categories\n",
    "ref.loc[(ref[\"structural_category\"] != \"antisense\") & (ref[\"structural_category\"] != \"fusion\") & (ref[\"structural_category\"] != \"intergenic\"), \"cat_assign\"]=\"one2one\"\n",
    "ref.loc[(ref[\"structural_category\"] == \"antisense\"), \"cat_assign\"]=\"antisense\"\n",
    "ref.loc[(ref[\"structural_category\"] == \"fusion\"), \"cat_assign\"]=\"fusion\"\n",
    "ref.loc[((ref[\"structural_category\"] == \"intergenic\") | (ref[\"structural_category\"] == \"genic_intron\")), \"cat_assign\"]=\"novel\"\n",
    "\n",
    "# in some cases, fusion transcripts were reported as one2one if they had perfect splice junctions over their fusion genes.\n",
    "# filter them and add replace the assignment\n",
    "ref[\"association\"] = ref[\"associated_gene\"].str.split(\"_\").str.len()\n",
    "ref.loc[(ref[\"cat_assign\"]==\"one2one\")&(ref[\"association\"]>1), \"cat_assign\"] = \"fusion\"\n",
    "\n",
    "# add gene read support\n",
    "# cat_read_support is the number of reads supporting the individual category assignments of a gene (can be different in terms of clear, fusion and antisense)\n",
    "# gene_association_read_support: sometimes there are multiple clear assignments within a gene, caused by a fusion gene also associated with the reads covering\n",
    "# multiple ref genes. This number gives the support for each clear-case gene in the reference.\n",
    "ref[\"cat_read_support\"]=ref.groupby(by=[\"cat_assign\",\"accession\",\"gene\"])[\"isoform_read_support\"].transform(\"sum\")\n",
    "ref[\"gene_association_read_support\"]=ref.groupby(by=[\"cat_assign\",\"accession\",\"gene\",\"associated_gene\"])[\"isoform_read_support\"].transform(\"sum\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73794c7c-826a-4a16-aaf9-dfcd5cdfde2c",
   "metadata": {},
   "source": [
    "## Prepare Fusion genes for both type assignment and later detailed analysis (fusion type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "426ad256-7645-47d2-8603-713b19b7f1a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def augustus_refs_fusion(df):\n",
    "    out_df= pd.DataFrame()\n",
    "    accessions=df[\"accession\"].unique().tolist()\n",
    "    for acc in accessions:    \n",
    "        busco = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/busco_{acc}.txt\", header=None)\n",
    "        house = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/housekeeping_{acc}.txt\", header=None)\n",
    "        nlr = pd.read_csv(f\"/ebio/abt6_projects7/dliso/dlis/leon_pipeline/analysis/sqanti/ref_genes/augustus_mappings/nlrs_{acc}.txt\", header=None)\n",
    "\n",
    "        busco[\"type\"] = \"busco\"\n",
    "        house[\"type\"] = \"housekeeping\"\n",
    "        nlr[\"type\"] = \"nlr\"\n",
    "        \n",
    "        gene_type = pd.concat([busco,house,nlr])\n",
    "        gene_type.columns = [\"ref_isoform\",\"type\"]\n",
    "        gene_type = gene_type.groupby(\"ref_isoform\")[\"type\"].apply(', '.join).reset_index()\n",
    "\n",
    "        gene_type[\"ref_gene\"] = gene_type[\"ref_isoform\"].str.split(\".\").str[0]\n",
    "        gene_type.loc[gene_type[\"type\"] == \"busco, nlr\"] = \"nlr\"\n",
    "        gene_type.loc[gene_type[\"type\"] == \"housekeeping, nlr\"] = \"nlr\"\n",
    "        gene_type.loc[gene_type[\"type\"] == \"busco, housekeeping\"] = \"housekeeping\"\n",
    "\n",
    "        sub=df.loc[df[\"accession\"]==acc]\n",
    "        sub = pd.merge(left=sub, right=gene_type, how='left', left_on='gene_split', right_on='ref_gene')\n",
    "        \n",
    "        out_df=pd.concat([out_df,sub])\n",
    "    return out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b7326ed2-2734-473c-be5a-245fe778e261",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fus=ref.loc[ref[\"cat_assign\"]==\"fusion\"].copy()\n",
    "fus[\"gene_split\"]=fus[\"associated_gene\"].str.split(\"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "32bb6aeb-25b7-441f-9a41-1d2d61672150",
   "metadata": {},
   "outputs": [],
   "source": [
    "split=fus.explode(\"gene_split\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8511ee37-7adf-4697-b227-782dde7da299",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test for novel genes in fusion genes\n",
    "for i in split[\"gene_split\"].tolist():\n",
    "    if i == \"novelGene\":\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9d12bf43-6337-47c4-ad0f-226fb62cb076",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fusion_types=augustus_refs_fusion(split).drop(columns=[\"ref_isoform\",\"ref_gene\"])\n",
    "fusion_types[\"type\"] = fusion_types[\"type\"].fillna(\"other\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1cdee804-e9ac-4c74-a1d1-d801e5e91695",
   "metadata": {},
   "outputs": [],
   "source": [
    "fusion_types.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "300cbe9c-ecae-44fe-b53d-c3b256bd7ee4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "fusion_out=pd.DataFrame()\n",
    "for i in fusion_types.groupby(by=[\"accession\", \"gene\", \"isoform\"], as_index=False):\n",
    "    t=i[1][\"type\"].value_counts()\n",
    "    t=t.reset_index()\n",
    "\n",
    "    t=t[\"index\"].tolist()\n",
    "\n",
    "    i[1][\"fusion_order\"] = \"_\".join(i[1][\"type\"].tolist())\n",
    "    \n",
    "    if \"busco\" in t:\n",
    "        i[1][\"busco\"] = 1\n",
    "    if \"nlr\" in t:\n",
    "        i[1][\"nlr\"] = 1\n",
    "    if \"housekeeping\" in t:\n",
    "        i[1][\"housekeeping\"] = 1\n",
    "    if \"other\" in t:\n",
    "        i[1][\"other\"] = 1\n",
    "\n",
    "    fusion_out=pd.concat([fusion_out,i[1]], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b1edb1e9-5a0e-4e19-a8d6-6988927291e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isoform</th>\n",
       "      <th>accession</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>exons</th>\n",
       "      <th>length</th>\n",
       "      <th>isoform_read_support</th>\n",
       "      <th>gene_read_support</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>...</th>\n",
       "      <th>association</th>\n",
       "      <th>cat_read_support</th>\n",
       "      <th>gene_association_read_support</th>\n",
       "      <th>gene_split</th>\n",
       "      <th>type</th>\n",
       "      <th>fusion_order</th>\n",
       "      <th>other</th>\n",
       "      <th>busco</th>\n",
       "      <th>nlr</th>\n",
       "      <th>housekeeping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2299</th>\n",
       "      <td>G10006</td>\n",
       "      <td>G10006.5</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g19413_g19414</td>\n",
       "      <td>novel</td>\n",
       "      <td>10</td>\n",
       "      <td>2461</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>g19413</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2300</th>\n",
       "      <td>G10006</td>\n",
       "      <td>G10006.5</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g19413_g19414</td>\n",
       "      <td>novel</td>\n",
       "      <td>10</td>\n",
       "      <td>2461</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>g19414</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2301</th>\n",
       "      <td>G10013</td>\n",
       "      <td>G10013.3</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g19424_g19425</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>3552</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>g19424</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2302</th>\n",
       "      <td>G10013</td>\n",
       "      <td>G10013.3</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g19424_g19425</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>3552</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>g19425</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2303</th>\n",
       "      <td>G10013</td>\n",
       "      <td>G10013.5</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g19424_g19425</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>3534</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>g19424</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72219</th>\n",
       "      <td>G9951</td>\n",
       "      <td>G9951.9</td>\n",
       "      <td>col0</td>\n",
       "      <td>AT4G00030_AT4G00040</td>\n",
       "      <td>AT4G00030.1</td>\n",
       "      <td>3</td>\n",
       "      <td>2332</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>full-splice_match</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>AT4G00040</td>\n",
       "      <td>other</td>\n",
       "      <td>busco_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72220</th>\n",
       "      <td>G9987</td>\n",
       "      <td>G9987.2</td>\n",
       "      <td>col0</td>\n",
       "      <td>AT4G00560_AT4G00570</td>\n",
       "      <td>novel</td>\n",
       "      <td>23</td>\n",
       "      <td>3761</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>AT4G00560</td>\n",
       "      <td>busco</td>\n",
       "      <td>busco_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72221</th>\n",
       "      <td>G9987</td>\n",
       "      <td>G9987.2</td>\n",
       "      <td>col0</td>\n",
       "      <td>AT4G00560_AT4G00570</td>\n",
       "      <td>novel</td>\n",
       "      <td>23</td>\n",
       "      <td>3761</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>AT4G00570</td>\n",
       "      <td>other</td>\n",
       "      <td>busco_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72222</th>\n",
       "      <td>G9989</td>\n",
       "      <td>G9989.11</td>\n",
       "      <td>col0</td>\n",
       "      <td>AT4G00630_AT4G00620</td>\n",
       "      <td>novel</td>\n",
       "      <td>8</td>\n",
       "      <td>1542</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>AT4G00630</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72223</th>\n",
       "      <td>G9989</td>\n",
       "      <td>G9989.11</td>\n",
       "      <td>col0</td>\n",
       "      <td>AT4G00630_AT4G00620</td>\n",
       "      <td>novel</td>\n",
       "      <td>8</td>\n",
       "      <td>1542</td>\n",
       "      <td>1</td>\n",
       "      <td>88</td>\n",
       "      <td>fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>AT4G00620</td>\n",
       "      <td>other</td>\n",
       "      <td>other_other</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73597 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene   isoform accession      associated_gene associated_transcript  \\\n",
       "2299   G10006  G10006.5    at6923        g19413_g19414                 novel   \n",
       "2300   G10006  G10006.5    at6923        g19413_g19414                 novel   \n",
       "2301   G10013  G10013.3    at6923        g19424_g19425                 novel   \n",
       "2302   G10013  G10013.3    at6923        g19424_g19425                 novel   \n",
       "2303   G10013  G10013.5    at6923        g19424_g19425                 novel   \n",
       "...       ...       ...       ...                  ...                   ...   \n",
       "72219   G9951   G9951.9      col0  AT4G00030_AT4G00040           AT4G00030.1   \n",
       "72220   G9987   G9987.2      col0  AT4G00560_AT4G00570                 novel   \n",
       "72221   G9987   G9987.2      col0  AT4G00560_AT4G00570                 novel   \n",
       "72222   G9989  G9989.11      col0  AT4G00630_AT4G00620                 novel   \n",
       "72223   G9989  G9989.11      col0  AT4G00630_AT4G00620                 novel   \n",
       "\n",
       "       exons  length  isoform_read_support  gene_read_support  \\\n",
       "2299      10    2461                     1                 11   \n",
       "2300      10    2461                     1                 11   \n",
       "2301       2    3552                     1                  5   \n",
       "2302       2    3552                     1                  5   \n",
       "2303       2    3534                     1                  5   \n",
       "...      ...     ...                   ...                ...   \n",
       "72219      3    2332                     1                 21   \n",
       "72220     23    3761                     1                 71   \n",
       "72221     23    3761                     1                 71   \n",
       "72222      8    1542                     1                 88   \n",
       "72223      8    1542                     1                 88   \n",
       "\n",
       "      structural_category  ... association cat_read_support  \\\n",
       "2299               fusion  ...           2                1   \n",
       "2300               fusion  ...           2                1   \n",
       "2301               fusion  ...           2                2   \n",
       "2302               fusion  ...           2                2   \n",
       "2303               fusion  ...           2                2   \n",
       "...                   ...  ...         ...              ...   \n",
       "72219   full-splice_match  ...           2               19   \n",
       "72220              fusion  ...           2                1   \n",
       "72221              fusion  ...           2                1   \n",
       "72222              fusion  ...           2                1   \n",
       "72223              fusion  ...           2                1   \n",
       "\n",
       "      gene_association_read_support  gene_split   type  fusion_order other  \\\n",
       "2299                              1      g19413  other   other_other   1.0   \n",
       "2300                              1      g19414  other   other_other   1.0   \n",
       "2301                              2      g19424  other   other_other   1.0   \n",
       "2302                              2      g19425  other   other_other   1.0   \n",
       "2303                              2      g19424  other   other_other   1.0   \n",
       "...                             ...         ...    ...           ...   ...   \n",
       "72219                            19   AT4G00040  other   busco_other   1.0   \n",
       "72220                             1   AT4G00560  busco   busco_other   1.0   \n",
       "72221                             1   AT4G00570  other   busco_other   1.0   \n",
       "72222                             1   AT4G00630  other   other_other   1.0   \n",
       "72223                             1   AT4G00620  other   other_other   1.0   \n",
       "\n",
       "      busco nlr  housekeeping  \n",
       "2299    NaN NaN           NaN  \n",
       "2300    NaN NaN           NaN  \n",
       "2301    NaN NaN           NaN  \n",
       "2302    NaN NaN           NaN  \n",
       "2303    NaN NaN           NaN  \n",
       "...     ...  ..           ...  \n",
       "72219   1.0 NaN           NaN  \n",
       "72220   1.0 NaN           NaN  \n",
       "72221   1.0 NaN           NaN  \n",
       "72222   NaN NaN           NaN  \n",
       "72223   NaN NaN           NaN  \n",
       "\n",
       "[73597 rows x 23 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fusion_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "92e4e4b5-9188-4654-ad14-920663873f94",
   "metadata": {},
   "outputs": [],
   "source": [
    "fusion_types= fusion_out.drop(columns=[\"gene_split\",\"type\",\"association\"]).drop_duplicates()\n",
    "fusion_types.fillna({'other':0, 'busco':0, 'nlr':0,'housekeeping':0}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d1e9df9-d4da-4175-a5b5-dcc78820035a",
   "metadata": {},
   "source": [
    "## Connect all ref gene sets (Clear Cases + Fusion genes (Order: NLR>House>Busco>Other)) - This creates the DF for subsequent analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad1c3dd4-f7d1-4458-b545-5a987197637d",
   "metadata": {},
   "source": [
    "### The reason to do it this preciesely for the reference groups to have a fair comparison between them and add duplicate fusion genes to each of the potential groups it belongs. In the antisense case, it was done quick without distinguishing and just keeping the order above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "39a867cb-d96f-4245-bc0d-b9292b480a2f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "NLRs=pd.DataFrame()\n",
    "for acc in ref[\"accession\"].unique().tolist():\n",
    "    \n",
    "    clear_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"cat_assign\"] == \"one2one\")]\n",
    "    isos_type=augustus_refs(clear_isos).drop(columns=[\"ref_isoform\",\"ref_gene\"])\n",
    "    nlr_isos=isos_type.loc[isos_type[\"type\"]==\"nlr\"]\n",
    "\n",
    "    \n",
    "    b=fusion_types.loc[(fusion_types[\"nlr\"]==1)&(fusion_types[\"accession\"]==acc)][\"isoform\"].tolist()\n",
    "    fusion_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"isoform\"].isin(b))]\n",
    "\n",
    "    c=pd.concat([nlr_isos,fusion_isos], ignore_index=True)\n",
    "    \n",
    "    c[\"isoforms_per_gene\"]= c.groupby(by=\"gene\")[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    c[\"ORFs_per_gene\"]= c.groupby(by=[\"gene\"])[\"Sqanti_ORF_seq\"].transform(lambda x: len(set(x.tolist())))\n",
    "    c[\"isoforms_per_ORFs\"]= c.groupby(by=[\"gene\",\"Sqanti_ORF_seq\"])[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    \n",
    "    #print(\"Mean nlr isoforms per gene:\")\n",
    "    hipg=c[[\"gene\",\"isoforms_per_gene\"]].drop_duplicates()\n",
    "    #print(hipg[\"isoforms_per_gene\"].apply(\"mean\"))\n",
    "\n",
    "    #print(\"Mean nlr ORFs per gene:\")\n",
    "    hopg=c[[\"gene\",\"ORFs_per_gene\"]].drop_duplicates()\n",
    "    #print(hopg[\"ORFs_per_gene\"].apply(\"mean\"))\n",
    "    #print(\"\\n\")\n",
    "    \n",
    "    \n",
    "    NLRs=pd.concat([NLRs,c])\n",
    "NLRs[\"type\"]=\"nlr\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "df45894e-9051-4a70-a030-a5e6d6088b0a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "House=pd.DataFrame()\n",
    "for acc in ref[\"accession\"].unique().tolist():\n",
    "    \n",
    "    clear_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"cat_assign\"] == \"one2one\")]\n",
    "    isos_type=augustus_refs(clear_isos).drop(columns=[\"ref_isoform\",\"ref_gene\"])\n",
    "    house_isos=isos_type.loc[isos_type[\"type\"]==\"housekeeping\"]\n",
    "\n",
    "    \n",
    "    b=fusion_types.loc[(fusion_types[\"housekeeping\"]==1)&(fusion_types[\"accession\"]==acc)][\"isoform\"].tolist()\n",
    "    fusion_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"isoform\"].isin(b))]\n",
    "\n",
    "    c=pd.concat([house_isos,fusion_isos], ignore_index=True)\n",
    "    #c=ref.loc[(ref[\"accession\"]==f\"{acc}\") & (ref[\"gene\"].isin(set(a[f\"{acc}\"].tolist() + b[f\"{acc}\"].tolist())))&~(ref[\"cat_assign\"].isin([\"antisense\",\"novel\"]))].copy()\n",
    "    \n",
    "    c[\"isoforms_per_gene\"]= c.groupby(by=\"gene\")[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    c[\"ORFs_per_gene\"]= c.groupby(by=[\"gene\"])[\"Sqanti_ORF_seq\"].transform(lambda x: len(set(x.tolist())))\n",
    "    c[\"isoforms_per_ORFs\"]= c.groupby(by=[\"gene\",\"Sqanti_ORF_seq\"])[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    \n",
    "    #print(\"Mean nlr isoforms per gene:\")\n",
    "    hipg=c[[\"gene\",\"isoforms_per_gene\"]].drop_duplicates()\n",
    "    #print(hipg[\"isoforms_per_gene\"].apply(\"mean\"))\n",
    "\n",
    "    #print(\"Mean nlr ORFs per gene:\")\n",
    "    hopg=c[[\"gene\",\"ORFs_per_gene\"]].drop_duplicates()\n",
    "    #print(hopg[\"ORFs_per_gene\"].apply(\"mean\"))\n",
    "    #print(\"\\n\")\n",
    "    \n",
    "    \n",
    "    House=pd.concat([House,c])\n",
    "House[\"type\"]=\"housekeeping\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f18ef2a-7759-44c8-8ca3-0d62b8c7a39e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Busco=pd.DataFrame()\n",
    "for acc in ref[\"accession\"].unique().tolist():\n",
    "    \n",
    "    clear_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"cat_assign\"] == \"one2one\")]\n",
    "    isos_type=augustus_refs(clear_isos).drop(columns=[\"ref_isoform\",\"ref_gene\"])\n",
    "    busco_isos=isos_type.loc[isos_type[\"type\"]==\"busco\"]\n",
    "\n",
    "    \n",
    "    b=fusion_types.loc[(fusion_types[\"busco\"]==1)&(fusion_types[\"accession\"]==acc)][\"isoform\"].tolist()\n",
    "    fusion_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"isoform\"].isin(b))]\n",
    "\n",
    "    c=pd.concat([busco_isos,fusion_isos], ignore_index=True)\n",
    "    #c=ref.loc[(ref[\"accession\"]==f\"{acc}\") & (ref[\"gene\"].isin(set(a[f\"{acc}\"].tolist() + b[f\"{acc}\"].tolist())))&~(ref[\"cat_assign\"].isin([\"antisense\",\"novel\"]))].copy()\n",
    "    \n",
    "    c[\"isoforms_per_gene\"]= c.groupby(by=\"gene\")[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    c[\"ORFs_per_gene\"]= c.groupby(by=[\"gene\"])[\"Sqanti_ORF_seq\"].transform(lambda x: len(set(x.tolist())))\n",
    "    c[\"isoforms_per_ORFs\"]= c.groupby(by=[\"gene\",\"Sqanti_ORF_seq\"])[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    \n",
    "    #print(\"Mean nlr isoforms per gene:\")\n",
    "    hipg=c[[\"gene\",\"isoforms_per_gene\"]].drop_duplicates()\n",
    "    #print(hipg[\"isoforms_per_gene\"].apply(\"mean\"))\n",
    "\n",
    "    #print(\"Mean nlr ORFs per gene:\")\n",
    "    hopg=c[[\"gene\",\"ORFs_per_gene\"]].drop_duplicates()\n",
    "    #print(hopg[\"ORFs_per_gene\"].apply(\"mean\"))\n",
    "    #print(\"\\n\")\n",
    "    \n",
    "    \n",
    "    Busco=pd.concat([Busco,c])\n",
    "Busco[\"type\"]=\"busco\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a2a2cfad-22f5-4264-8665-6d8299c3cc73",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Other=pd.DataFrame()\n",
    "for acc in ref[\"accession\"].unique().tolist():\n",
    "    \n",
    "    clear_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"cat_assign\"] == \"one2one\")]\n",
    "    isos_type=augustus_refs(clear_isos).drop(columns=[\"ref_isoform\",\"ref_gene\"])\n",
    "    isos_type[\"type\"] = isos_type[\"type\"].fillna(\"other\")\n",
    "    other_isos=isos_type.loc[isos_type[\"type\"]==\"other\"]\n",
    "\n",
    "    \n",
    "    b=fusion_types.loc[(fusion_types[\"other\"]==1)&(fusion_types[\"accession\"]==acc)][\"isoform\"].tolist()\n",
    "    fusion_isos=ref.loc[(ref[\"accession\"]==acc)&(ref[\"isoform\"].isin(b))]\n",
    "\n",
    "    c=pd.concat([other_isos,fusion_isos], ignore_index=True)\n",
    "    \n",
    "    c[\"isoforms_per_gene\"]= c.groupby(by=\"gene\")[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    c[\"ORFs_per_gene\"]= c.groupby(by=[\"gene\"])[\"Sqanti_ORF_seq\"].transform(lambda x: len(set(x.tolist())))\n",
    "    c[\"isoforms_per_ORFs\"]= c.groupby(by=[\"gene\",\"Sqanti_ORF_seq\"])[\"isoform\"].transform(lambda x: len(x.tolist()))\n",
    "    \n",
    "    #print(\"Mean nlr isoforms per gene:\")\n",
    "    hipg=c[[\"gene\",\"isoforms_per_gene\"]].drop_duplicates()\n",
    "    #print(hipg[\"isoforms_per_gene\"].apply(\"mean\"))\n",
    "\n",
    "    #print(\"Mean nlr ORFs per gene:\")\n",
    "    hopg=c[[\"gene\",\"ORFs_per_gene\"]].drop_duplicates()\n",
    "    #print(hopg[\"ORFs_per_gene\"].apply(\"mean\"))\n",
    "    #print(\"\\n\")\n",
    "    \n",
    "    \n",
    "    Other=pd.concat([Other,c])\n",
    "Other[\"type\"]=\"other\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d796da52-2f37-43eb-9382-74a72f8d6479",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isoform</th>\n",
       "      <th>accession</th>\n",
       "      <th>associated_gene</th>\n",
       "      <th>associated_transcript</th>\n",
       "      <th>exons</th>\n",
       "      <th>length</th>\n",
       "      <th>isoform_read_support</th>\n",
       "      <th>gene_read_support</th>\n",
       "      <th>structural_category</th>\n",
       "      <th>subcategory</th>\n",
       "      <th>Sqanti_ORF_seq</th>\n",
       "      <th>cat_assign</th>\n",
       "      <th>association</th>\n",
       "      <th>cat_read_support</th>\n",
       "      <th>gene_association_read_support</th>\n",
       "      <th>type</th>\n",
       "      <th>isoforms_per_gene</th>\n",
       "      <th>ORFs_per_gene</th>\n",
       "      <th>isoforms_per_ORFs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1128</td>\n",
       "      <td>G1128.10</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g1555</td>\n",
       "      <td>novel</td>\n",
       "      <td>1</td>\n",
       "      <td>3275</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>genic</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>MGNCVSLDVSCHHQTLNHACGCLFGDGNYIHMMEANLEALEKTIQE...</td>\n",
       "      <td>one2one</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>12</td>\n",
       "      <td>nlr</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1128</td>\n",
       "      <td>G1128.11</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g1555</td>\n",
       "      <td>novel</td>\n",
       "      <td>1</td>\n",
       "      <td>2744</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>genic</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...</td>\n",
       "      <td>one2one</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>12</td>\n",
       "      <td>nlr</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G1128</td>\n",
       "      <td>G1128.12</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g1555</td>\n",
       "      <td>novel</td>\n",
       "      <td>1</td>\n",
       "      <td>2795</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>genic</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...</td>\n",
       "      <td>one2one</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>12</td>\n",
       "      <td>nlr</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G1128</td>\n",
       "      <td>G1128.13</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g1555</td>\n",
       "      <td>novel</td>\n",
       "      <td>1</td>\n",
       "      <td>1821</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>genic</td>\n",
       "      <td>mono-exon</td>\n",
       "      <td>MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...</td>\n",
       "      <td>one2one</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>12</td>\n",
       "      <td>nlr</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G1128</td>\n",
       "      <td>G1128.2</td>\n",
       "      <td>at6923</td>\n",
       "      <td>g1555</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>3331</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>novel_not_in_catalog</td>\n",
       "      <td>intron_retention</td>\n",
       "      <td>MGNCVSLDVSCHHQTLNHACGCLFGDGNYIHMMEANLEALEKTIQE...</td>\n",
       "      <td>one2one</td>\n",
       "      <td>1</td>\n",
       "      <td>59</td>\n",
       "      <td>12</td>\n",
       "      <td>nlr</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3419863</th>\n",
       "      <td>G16302</td>\n",
       "      <td>G16302.24</td>\n",
       "      <td>col0</td>\n",
       "      <td>ATCG01310_ATCG01300</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>818</td>\n",
       "      <td>1</td>\n",
       "      <td>136</td>\n",
       "      <td>fusion</td>\n",
       "      <td>multi-exon</td>\n",
       "      <td>MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>other</td>\n",
       "      <td>49</td>\n",
       "      <td>8</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3419864</th>\n",
       "      <td>G16302</td>\n",
       "      <td>G16302.26</td>\n",
       "      <td>col0</td>\n",
       "      <td>ATCG01310_ATCG01300</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>831</td>\n",
       "      <td>2</td>\n",
       "      <td>136</td>\n",
       "      <td>fusion</td>\n",
       "      <td>multi-exon</td>\n",
       "      <td>MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>other</td>\n",
       "      <td>49</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3419865</th>\n",
       "      <td>G16302</td>\n",
       "      <td>G16302.5</td>\n",
       "      <td>col0</td>\n",
       "      <td>ATCG01310_ATCG01300</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>1204</td>\n",
       "      <td>1</td>\n",
       "      <td>136</td>\n",
       "      <td>fusion</td>\n",
       "      <td>multi-exon</td>\n",
       "      <td>MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>other</td>\n",
       "      <td>49</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3419866</th>\n",
       "      <td>G16302</td>\n",
       "      <td>G16302.7</td>\n",
       "      <td>col0</td>\n",
       "      <td>ATCG01310_ATCG01300</td>\n",
       "      <td>novel</td>\n",
       "      <td>2</td>\n",
       "      <td>1141</td>\n",
       "      <td>6</td>\n",
       "      <td>136</td>\n",
       "      <td>fusion</td>\n",
       "      <td>multi-exon</td>\n",
       "      <td>MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...</td>\n",
       "      <td>fusion</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>other</td>\n",
       "      <td>49</td>\n",
       "      <td>8</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3419867</th>\n",
       "      <td>G16227</td>\n",
       "      <td>G16227.1</td>\n",
       "      <td>col0</td>\n",
       "      <td>ATMG01130_ATMG00516_ATMG01120</td>\n",
       "      <td>novel</td>\n",
       "      <td>3</td>\n",
       "      <td>2363</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>fusion</td>\n",
       "      <td>multi-exon</td>\n",
       "      <td>MKEAIRMVLESIYDPEFPDTSHFRSGQGCHSVLRRIKEEWGISRWF...</td>\n",
       "      <td>fusion</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>other</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3419868 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene    isoform accession                associated_gene  \\\n",
       "0         G1128   G1128.10    at6923                          g1555   \n",
       "1         G1128   G1128.11    at6923                          g1555   \n",
       "2         G1128   G1128.12    at6923                          g1555   \n",
       "3         G1128   G1128.13    at6923                          g1555   \n",
       "4         G1128    G1128.2    at6923                          g1555   \n",
       "...         ...        ...       ...                            ...   \n",
       "3419863  G16302  G16302.24      col0            ATCG01310_ATCG01300   \n",
       "3419864  G16302  G16302.26      col0            ATCG01310_ATCG01300   \n",
       "3419865  G16302   G16302.5      col0            ATCG01310_ATCG01300   \n",
       "3419866  G16302   G16302.7      col0            ATCG01310_ATCG01300   \n",
       "3419867  G16227   G16227.1      col0  ATMG01130_ATMG00516_ATMG01120   \n",
       "\n",
       "        associated_transcript  exons  length  isoform_read_support  \\\n",
       "0                       novel      1    3275                     1   \n",
       "1                       novel      1    2744                     1   \n",
       "2                       novel      1    2795                     1   \n",
       "3                       novel      1    1821                     1   \n",
       "4                       novel      2    3331                     1   \n",
       "...                       ...    ...     ...                   ...   \n",
       "3419863                 novel      2     818                     1   \n",
       "3419864                 novel      2     831                     2   \n",
       "3419865                 novel      2    1204                     1   \n",
       "3419866                 novel      2    1141                     6   \n",
       "3419867                 novel      3    2363                     1   \n",
       "\n",
       "         gene_read_support   structural_category       subcategory  \\\n",
       "0                       59                 genic         mono-exon   \n",
       "1                       59                 genic         mono-exon   \n",
       "2                       59                 genic         mono-exon   \n",
       "3                       59                 genic         mono-exon   \n",
       "4                       59  novel_not_in_catalog  intron_retention   \n",
       "...                    ...                   ...               ...   \n",
       "3419863                136                fusion        multi-exon   \n",
       "3419864                136                fusion        multi-exon   \n",
       "3419865                136                fusion        multi-exon   \n",
       "3419866                136                fusion        multi-exon   \n",
       "3419867                  3                fusion        multi-exon   \n",
       "\n",
       "                                            Sqanti_ORF_seq cat_assign  \\\n",
       "0        MGNCVSLDVSCHHQTLNHACGCLFGDGNYIHMMEANLEALEKTIQE...    one2one   \n",
       "1        MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...    one2one   \n",
       "2        MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...    one2one   \n",
       "3        MDGVLTFTVKMHDVIREMALWIASKFGKQKETLCVRSGAQLRQIPK...    one2one   \n",
       "4        MGNCVSLDVSCHHQTLNHACGCLFGDGNYIHMMEANLEALEKTIQE...    one2one   \n",
       "...                                                    ...        ...   \n",
       "3419863  MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...     fusion   \n",
       "3419864  MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...     fusion   \n",
       "3419865  MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...     fusion   \n",
       "3419866  MAIHLYKTSTPSTRNGAVDSQVKSNPRNNLICGQHHCGKGRNARGI...     fusion   \n",
       "3419867  MKEAIRMVLESIYDPEFPDTSHFRSGQGCHSVLRRIKEEWGISRWF...     fusion   \n",
       "\n",
       "         association  cat_read_support  gene_association_read_support   type  \\\n",
       "0                  1                59                             12    nlr   \n",
       "1                  1                59                             12    nlr   \n",
       "2                  1                59                             12    nlr   \n",
       "3                  1                59                             12    nlr   \n",
       "4                  1                59                             12    nlr   \n",
       "...              ...               ...                            ...    ...   \n",
       "3419863            2                15                             15  other   \n",
       "3419864            2                15                             15  other   \n",
       "3419865            2                15                             15  other   \n",
       "3419866            2                15                             15  other   \n",
       "3419867            3                 1                              1  other   \n",
       "\n",
       "         isoforms_per_gene  ORFs_per_gene  isoforms_per_ORFs  \n",
       "0                       11              2                8.0  \n",
       "1                       11              2                3.0  \n",
       "2                       11              2                3.0  \n",
       "3                       11              2                3.0  \n",
       "4                       11              2                8.0  \n",
       "...                    ...            ...                ...  \n",
       "3419863                 49              8                2.0  \n",
       "3419864                 49              8                4.0  \n",
       "3419865                 49              8                4.0  \n",
       "3419866                 49              8                4.0  \n",
       "3419867                  3              2                1.0  \n",
       "\n",
       "[3419868 rows x 20 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_types=pd.concat([NLRs,Busco,House,Other], ignore_index=True)\n",
    "all_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f66ae28f-d12a-48d1-b161-1983d4634954",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 18/18 [03:30<00:00, 11.72s/it]\n"
     ]
    }
   ],
   "source": [
    "drop=[]\n",
    "for i in tqdm(all_types.groupby(\"accession\")):\n",
    "    df = i[1]\n",
    "    boolean=df[\"isoform\"].value_counts()>1\n",
    "    boolean=boolean.rename_axis('isoform').reset_index(name='bool')\n",
    "    #print(boolean.loc[boolean[\"bool\"]==True])\n",
    "    #print(boolean)\n",
    "    dups = boolean.loc[boolean[\"bool\"]==True][\"isoform\"].tolist()\n",
    "    for iso in dups:\n",
    "        iso_df=df[df[\"isoform\"]==iso]\n",
    "        if \"nlr\" in iso_df[\"type\"].tolist():\n",
    "            drop.extend(iso_df[~(iso_df[\"type\"]==\"nlr\")].index.tolist())\n",
    "            continue\n",
    "        if \"housekeeping\" in iso_df[\"type\"].tolist():\n",
    "            drop.extend(iso_df[~(iso_df[\"type\"]==\"housekeeping\")].index.tolist())\n",
    "            continue\n",
    "        if \"busco\" in iso_df[\"type\"].tolist():\n",
    "            drop.extend(iso_df[~(iso_df[\"type\"]==\"busco\")].index.tolist())\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a2d0a0af-162a-4890-b35c-e63c0ad47dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_types=all_types.drop(drop).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "34be67d2-4ea0-4550-8ea1-aaf01e3db6e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is the fall_typesdataframe to continue \n",
    "all_types\n",
    "all_types.to_pickle(\"clear-fusion/clear-fusion_tama_augustus.pkl\")\n",
    "#all_types.to_pickle(\"../5_NLRs/clear-fusion_tama_augustus.pkl\")\n",
    "#all_types.to_pickle(\"../3_fusion_genes/clear-fusion_tama_augustus.pkl\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
